File: typescript/tsfind/src/finder.ts
/*
 * finder.ts
 *
 * performs the finding based on the given FindSettings instance
 */

'use strict';

import * as assert from 'assert';
import * as fs from 'fs';
import {stat, lstat} from 'fs/promises';
import * as path from 'path';

import {FileResult} from './fileresult';
import {FileResultFormatter} from "./fileresultformatter";
import {FileResultSorter} from "./fileresultsorter";
import {FileType} from './filetype';
import {FileTypes} from './filetypes';
import {FileUtil, ENOENT, EACCES} from './fileutil';
import {FindError} from './finderror';
import {FindSettings} from './findsettings';
import * as common from "./common";

const startPathNotDefined = 'Startpath not defined';
const invalidRangeForMinDepthAndMaxDepth = 'Invalid range for mindepth and maxdepth';
const invalidRangeForMinLastModAndMaxLastMod = 'Invalid range for minlastmod and maxlastmod';
const invalidRangeForMinSizeAndMaxSize = 'Invalid range for minsize and maxsize';
const startPathNotFound = 'Startpath not found';
const startPathNotReadable = 'Startpath not readable';
const startPathDoesNotMatchFindSettings = 'Startpath does not match find settings';


export class Finder {
    _settings: FindSettings;

    constructor(settings: FindSettings) {
        this._settings = settings;
        this.validateSettings();
    }

    private validateSettings(): void {
        try {
            assert.ok(this._settings.paths.length > 0, startPathNotDefined);
            for (let p of this._settings.paths) {
                // Validate existence, accessibility and "findability" of file path (directory or regular file)
                try {
                    fs.accessSync(p, fs.constants.F_OK | fs.constants.R_OK);
                } catch {
                    p = FileUtil.expandPath(p);
                    fs.accessSync(p, fs.constants.F_OK | fs.constants.R_OK);
                }
                let stats = fs.lstatSync(p);
                if (stats.isSymbolicLink()) {
                    assert.ok(this._settings.followSymlinks, startPathDoesNotMatchFindSettings);
                }
                stats = fs.statSync(p);
                if (stats.isDirectory()) {
                    assert.ok(this.isTraversableDirPath(p), startPathDoesNotMatchFindSettings);
                } else if (stats.isFile()) {
                    assert.ok(this.filterToFileResult(p, stats) !== null,
                        startPathDoesNotMatchFindSettings);
                } else {
                    // TODO: start path is unknown/invalid type
                    throw new FindError(startPathDoesNotMatchFindSettings);
                }
            }
            if (this._settings.maxDepth > -1 && this._settings.minDepth > -1) {
                assert.ok(this._settings.maxDepth >= this._settings.minDepth,
                    invalidRangeForMinDepthAndMaxDepth);
            }
            if (this._settings.maxLastMod > 0 && this._settings.minLastMod > 0) {
                assert.ok(this._settings.maxLastMod >= this._settings.minLastMod,
                    invalidRangeForMinLastModAndMaxLastMod);
            }
            if (this._settings.maxSize > 0 && this._settings.minSize > 0) {
                assert.ok(this._settings.maxSize >= this._settings.minSize,
                    invalidRangeForMinSizeAndMaxSize);
            }

        } catch (err: Error | any) {
            let msg = err.message;
            if (err.code === ENOENT) {
                msg = startPathNotFound;
            } else if (err.code === EACCES) {
                msg = startPathNotReadable;
            }
            throw new FindError(msg);
        }
    }

    public static matchesAnyPattern(s: string, patterns: RegExp[]): boolean {
        return patterns.some((p: RegExp) => s.search(p) > -1);
    }

    public static anyMatchesAnyPattern(ss: string[], patterns: RegExp[]): boolean {
        return ss.some((s: string) => Finder.matchesAnyPattern(s, patterns));
    }

    public static matchesAnyString(s: string, elements: string[]): boolean {
        return elements.indexOf(s) > -1;
    }

    public static matchesAnyFileType(ft: FileType, fileTypes: FileType[]): boolean {
        return fileTypes.indexOf(ft) > -1;
    }

    public static emptyOrMatchesAnyPattern(s: string, patterns: RegExp[]): boolean {
        return patterns.length === 0 || Finder.matchesAnyPattern(s, patterns);
    }

    public static emptyOrNotMatchesAnyPattern(s: string, patterns: RegExp[]): boolean {
        return patterns.length === 0 || !Finder.matchesAnyPattern(s, patterns);
    }

    public static emptyOrAnyMatchesAnyPattern(ss: string[], patterns: RegExp[]): boolean {
        return ss.some((s: string) => Finder.emptyOrMatchesAnyPattern(s, patterns));
    }

    public static emptyOrNotAnyMatchesAnyPattern(ss: string[], patterns: RegExp[]): boolean {
        return ss.some((s: string) => !Finder.emptyOrMatchesAnyPattern(s, patterns));
    }

    public static emptyOrMatchesAnyString(s: string, elements: string[]): boolean {
        return elements.length === 0 || Finder.matchesAnyString(s, elements);
    }

    public static emptyOrNotMatchesAnyString(s: string, elements: string[]): boolean {
        return elements.length === 0 || !Finder.matchesAnyString(s, elements);
    }

    public static emptyOrMatchesAnyFileType(ft: FileType, fileTypes: FileType[]): boolean {
        return fileTypes.length === 0 || Finder.matchesAnyFileType(ft, fileTypes);
    }

    public static emptyOrNotMatchesAnyFileType(ft: FileType, fileTypes: FileType[]): boolean {
        return fileTypes.length === 0 || !Finder.matchesAnyFileType(ft, fileTypes);
    }

    public isMatchingDirPathByHidden(dirPath: string): boolean {
        return this._settings.includeHidden || !FileUtil.isHiddenPath(dirPath);
    }

    public isMatchingDirPathByInPatterns(dirPath: string): boolean {
        const elems = FileUtil.getPathElems(dirPath);
        return Finder.emptyOrAnyMatchesAnyPattern(elems, this._settings.inDirPatterns);
    }

    public isMatchingDirPathByOutPatterns(dirPath: string): boolean {
        const elems = FileUtil.getPathElems(dirPath);
        return Finder.emptyOrNotAnyMatchesAnyPattern(elems, this._settings.outDirPatterns);
    }

    public isTraversableDirPath(dirPath: string): boolean {
        return this.isMatchingDirPathByHidden(dirPath)
            && this.isMatchingDirPathByOutPatterns(dirPath);
    }

    public isMatchingDirPath(dirPath: string): boolean {
        return this.isMatchingDirPathByHidden(dirPath)
            && this.isMatchingDirPathByInPatterns(dirPath)
            && this.isMatchingDirPathByOutPatterns(dirPath);
    }

    public isNullOrMatchingDirPath(dirPath: string): boolean {
        if (!dirPath) return true;
        return this.isMatchingDirPath(dirPath);
    }

    public isMatchingFileNameByHidden(fileName: string): boolean {
        return this._settings.includeHidden || !FileUtil.isHiddenName(fileName);
    }

    public isMatchingArchiveExtension(ext: string): boolean {
        return (Finder.emptyOrMatchesAnyString(ext, this._settings.inArchiveExtensions)
            && Finder.emptyOrNotMatchesAnyString(ext, this._settings.outArchiveExtensions));
    }

    public isMatchingArchiveExtensionForFilePath(filePath: string): boolean {
        if (this._settings.inArchiveExtensions.length || this._settings.outArchiveExtensions.length) {
            const ext: string = FileUtil.getExtension(filePath);
            return this.isMatchingArchiveExtension(ext);
        }
        return true;
    }

    public isMatchingArchiveFileName(fileName: string): boolean {
        return (Finder.emptyOrMatchesAnyPattern(fileName, this._settings.inArchiveFilePatterns)
            && Finder.emptyOrNotMatchesAnyPattern(fileName, this._settings.outArchiveFilePatterns));
    }

    public isMatchingArchiveFileNameForFilePath(filePath: string): boolean {
        if (this._settings.inArchiveFilePatterns.length || this._settings.outArchiveFilePatterns.length) {
            return this.isMatchingArchiveFileName(path.basename(filePath));
        }
        return true;
    }

    public isMatchingArchiveFilePath(filePath: string): boolean {
        return this.isMatchingArchiveExtensionForFilePath(filePath)
            && this.isMatchingArchiveFileNameForFilePath(filePath);
    }

    public isMatchingArchiveFileResult(fr: FileResult): boolean {
        return this.isMatchingArchiveFilePath(fr.filePath);
    }

    public isMatchingExtension(ext: string): boolean {
        return (Finder.emptyOrMatchesAnyString(ext, this._settings.inExtensions)
            && Finder.emptyOrNotMatchesAnyString(ext, this._settings.outExtensions));
    }

    public isMatchingExtensionForFilePath(filePath: string): boolean {
        if (this._settings.inExtensions.length || this._settings.outExtensions.length) {
            const ext: string = FileUtil.getExtension(filePath);
            return this.isMatchingExtension(ext);
        }
        return true;
    }

    public isMatchingFileName(fileName: string): boolean {
        return (Finder.emptyOrMatchesAnyPattern(fileName, this._settings.inFilePatterns)
            && Finder.emptyOrNotMatchesAnyPattern(fileName, this._settings.outFilePatterns));
    }

    public isMatchingFileNameForFilePath(filePath: string): boolean {
        if (this._settings.inFilePatterns.length || this._settings.outFilePatterns.length) {
            return this.isMatchingFileName(path.basename(filePath));
        }
        return true;
    }

    public isMatchingFilePath(filePath: string): boolean {
        return (this.isMatchingExtensionForFilePath(filePath)
            && this.isMatchingFileNameForFilePath(filePath));
    }

    public isMatchingFileType(fileType: FileType): boolean {
        return (Finder.emptyOrMatchesAnyFileType(fileType, this._settings.inFileTypes)
            && Finder.emptyOrNotMatchesAnyFileType(fileType, this._settings.outFileTypes));
    }

    public isMatchingFileSize(fileSize: number): boolean {
        return ((this._settings.maxSize === 0 || fileSize <= this._settings.maxSize) &&
            (this._settings.minSize === 0 || fileSize >= this._settings.minSize));
    }

    public isMatchingLastMod(lastMod: number): boolean {
        return ((this._settings.maxLastMod === 0 || lastMod <= this._settings.maxLastMod) &&
            (this._settings.minLastMod === 0 || lastMod >= this._settings.minLastMod));
    }

    public isMatchingFileResult(fr: FileResult): boolean {
        return this.isMatchingFilePath(fr.filePath)
            && this.isMatchingFileType(fr.fileType)
            && this.isMatchingFileSize(fr.fileSize)
            && this.isMatchingLastMod(fr.lastMod);
    }

    public filterArchiveFilePathToFileResult(filePath: string): FileResult | null {
        if (!this._settings.includeArchives && !this._settings.archivesOnly) {
            return null;
        }

        if (!this.isMatchingArchiveFilePath(filePath)) {
            return null;
        }

        return new FileResult(filePath, FileType.Archive, 0, 0);
    }

    public filterRegularFilePathToFileResult(filePath: string, fileType: FileType, stat: fs.Stats | null = null): FileResult | null {
        if (this._settings.archivesOnly) {
            return null;
        }

        if (!this.isMatchingFilePath(filePath) || !this.isMatchingFileType(fileType)) {
            return null;
        }

        let fileSize = 0;
        let lastMod = 0;
        if (this._settings.needLastMod() || this._settings.needSize()) {
            stat = stat || fs.statSync(filePath);
            if (this._settings.needSize()) fileSize = stat.size;
            if (this._settings.needLastMod()) lastMod = stat.mtime.getTime();

            if (!this.isMatchingFileSize(fileSize) || !this.isMatchingLastMod(lastMod)) {
                return null;
            }
        }

        return new FileResult(filePath, fileType, fileSize, lastMod);
    }

    public filterToFileResult(filePath: string, stat: fs.Stats | null = null): FileResult | null {
        if (!this.isNullOrMatchingDirPath(path.dirname(filePath))
            || !this.isMatchingFileNameByHidden(path.basename(filePath))) {
            return null;
        }

        const fileType = FileTypes.getFileType(path.basename(filePath));
        if (fileType === FileType.Archive) {
            return this.filterArchiveFilePathToFileResult(filePath);
        }
        return this.filterRegularFilePathToFileResult(filePath, fileType, stat);
    }

    private async recGetFileResults(currentDir: string, minDepth: number, maxDepth: number, currentDepth: number): Promise<FileResult[]> {
        let fileResults: FileResult[] = [];
        let recurse: boolean = true;
        if (currentDepth === maxDepth) {
            recurse = false;
        } else if (maxDepth > -1 && currentDepth > maxDepth) {
            return [];
        }
        const findDirs: string[] = [];
        const filePaths = fs.readdirSync(currentDir, { recursive: false })
            .map(f =>  path.join(currentDir, f.toString()));
        for (const filePath of filePaths) {
            let stats = fs.lstatSync(filePath);
            if (stats.isSymbolicLink() && !this._settings.followSymlinks) {
                continue;
            }
            stats = fs.statSync(filePath);
            if (stats.isDirectory() && recurse && this.isTraversableDirPath(filePath)) {
                findDirs.push(filePath);
            } else if (stats.isFile() && (minDepth < 0 || currentDepth >= minDepth)) {
                const fr = this.filterToFileResult(filePath, stats);
                if (fr !== null) {
                    fileResults.push(fr);
                }
            }
        }
        const subDirFileResultArrays = await Promise.all(findDirs.map(d => this.recGetFileResults(d, minDepth, maxDepth, currentDepth + 1)));
        subDirFileResultArrays.forEach(subDirFileResults => {
            fileResults = fileResults.concat(subDirFileResults);
        });
        return fileResults;
    }

    private async getFileResults(filePath: string): Promise<FileResult[]> {
        try {
            fs.accessSync(filePath, fs.constants.F_OK | fs.constants.R_OK);
        } catch {
            filePath = FileUtil.expandPath(filePath);
        }
        let stats = await lstat(filePath);
        if (stats.isSymbolicLink() && !this._settings.followSymlinks) {
            throw new FindError(startPathDoesNotMatchFindSettings);
        }
        stats = await stat(filePath);
        if (stats.isDirectory()) {
            // if max_depth is zero, we can skip since a directory cannot be a result
            if (this._settings.maxDepth === 0) {
                return [];
            }
            if (this.isTraversableDirPath(filePath)) {
                let maxDepth = this._settings.maxDepth;
                if (!this._settings.recursive) {
                    maxDepth = 1;
                }
                return await this.recGetFileResults(filePath, this._settings.minDepth, maxDepth, 1);
            } else {
                throw new FindError(startPathDoesNotMatchFindSettings);
            }
        } else if (stats.isFile()) {
            // if min_depth > zero, we can skip since the file is at depth zero
            if (this._settings.minDepth > 0) {
                return [];
            }
            const fr = this.filterToFileResult(filePath);
            if (fr !== null) {
                return [fr];
            } else {
                throw new FindError(startPathDoesNotMatchFindSettings);
            }
        } else {
            throw new FindError(startPathDoesNotMatchFindSettings);
        }
    }

    public async find(): Promise<FileResult[]> {
        // get the file results
        let fileResults: FileResult[] = [];

        const pathFileResultsArrays =
            await Promise.all(this._settings.paths.map(p => this.getFileResults(p)));
        pathFileResultsArrays.forEach(pathFileResults => {
            fileResults = fileResults.concat(pathFileResults);
        });

        const fileResultSorter = new FileResultSorter(this._settings);
        fileResultSorter.sort(fileResults);
        return fileResults;
    }

    getMatchingDirs(fileResults: FileResult[]): string[] {
        const dirs: string[] = fileResults.map(fr => path.dirname(fr.filePath));
        return common.setFromArray(dirs);
    }

    printMatchingDirs(fileResults: FileResult[], formatter: FileResultFormatter): void {
        const dirs: string[] = this.getMatchingDirs(fileResults);
        if (dirs.length > 0) {
            common.log("\nMatching directories " + `(${dirs.length}):`);
            dirs.forEach(d => common.log(formatter.formatDirPath(d)));
        } else {
            common.log("\nMatching directories: 0");
        }
    }

    printMatchingFiles(fileResults: FileResult[], formatter: FileResultFormatter): void {
        if (fileResults.length > 0) {
            common.log("\nMatching files " + `(${fileResults.length}):`);
            fileResults.forEach(f => common.log(formatter.formatFileResult(f)));
        } else {
            common.log("\nMatching files: 0");
        }
    }
}

File: swift/swiftfind/Sources/swiftfind/Finder.swift
//
//  Finder.swift
//  swiftfind
//
//  Created by Cary Clark on 5/20/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

import Foundation
// FIXME: comparison operators with optionals were removed from the Swift Standard Libary.
// Consider refactoring the code to use the non-optional operators.
private func < <T: Comparable>(lhs: T?, rhs: T?) -> Bool {
    switch (lhs, rhs) {
    case let (l?, r?):
        l < r
    case (nil, _?):
        true
    default:
        false
    }
}

public let INVALID_RANGE_MINDEPTH_MAXDEPTH = "Invalid range for mindepth and maxdepth"
public let INVALID_RANGE_MINLASTMOD_MAXLASTMOD = "Invalid range for minlastmod and maxlastmod"
public let INVALID_RANGE_MINSIZE_MAXSIZE = "Invalid range for minsize and maxsize"
public let STARTPATH_NOT_DEFINED = "Startpath not defined"
public let STARTPATH_NOT_FOUND = "Startpath not found"
public let STARTPATH_NOT_READABLE = "Startpath not readable"
public let STARTPATH_NOT_MATCH_FIND_SETTINGS = "Startpath does not match find settings"

public class Finder {
    let fileTypes = FileTypes()
    let settings: FindSettings

    public init(settings: FindSettings) throws {
        self.settings = settings
        try validateSettings()
    }

    private func validateSettings() throws {
        if settings.paths.isEmpty {
            throw FindError(msg: STARTPATH_NOT_DEFINED)
        }
        for path in settings.paths {
            var p = path
            if (!FileUtil.exists(p)) {
                p = FileUtil.expandPath(p)
                if (!FileUtil.exists(p)) {
                    throw FindError(msg: STARTPATH_NOT_FOUND)
                }
            }
            if (!FileUtil.isReadableFile(p)) {
                throw FindError(msg: STARTPATH_NOT_READABLE)
            }

            var resolvedPath: String? = p
            if FileUtil.isSymlink(p) {
                if settings.followSymlinks {
                    resolvedPath = FileUtil.getSymlinkTarget(p)
                } else {
                    throw FindError(msg: STARTPATH_NOT_MATCH_FIND_SETTINGS)
                }
            }
            if FileUtil.isDirectory(resolvedPath!) {
                // still check p and not resolvedPath because p is the name entered
                if !isTraversableDirPath(p) {
                    throw FindError(msg: STARTPATH_NOT_MATCH_FIND_SETTINGS)
                }
            } else if FileUtil.isReadableFile(resolvedPath!) {
                // still check p and not resolvedPath because p is the name entered
                if filterToFileResult(p) == nil {
                    throw FindError(msg: STARTPATH_NOT_MATCH_FIND_SETTINGS)
                }
            } else {
                // TODO: start path is unknown/invalid type
                throw FindError(msg: STARTPATH_NOT_MATCH_FIND_SETTINGS)
            }
        }
        if settings.maxDepth > -1, settings.maxDepth < settings.minDepth {
            throw FindError(msg: INVALID_RANGE_MINDEPTH_MAXDEPTH)
        } else if settings.maxLastMod != nil, settings.minLastMod != nil, settings.maxLastMod! < settings.minLastMod! {
            throw FindError(msg: INVALID_RANGE_MINLASTMOD_MAXLASTMOD)
        } else if settings.maxSize > 0, settings.maxSize < settings.minSize {
            throw FindError(msg: INVALID_RANGE_MINSIZE_MAXSIZE)
        }
    }

    private func matchesAnyPattern(_ str: String, _ patterns: [Regex]) -> Bool {
        patterns.filter { $0.test(str) }.count > 0
    }

    private func anyMatchesAnyPattern(_ strs: [String], _ patterns: [Regex]) -> Bool {
        strs.filter { self.matchesAnyPattern($0, patterns) }.count > 0
    }

    private func emptyOrMatchesAnyPattern(_ str: String, _ patterns: [Regex]) -> Bool {
        patterns.isEmpty || matchesAnyPattern(str, patterns)
    }

    private func emptyOrNotMatchesAnyPattern(_ str: String, _ patterns: [Regex]) -> Bool {
        patterns.isEmpty || !matchesAnyPattern(str, patterns)
    }

    private func emptyOrAnyMatchesAnyPattern(_ strs: [String], _ patterns: [Regex]) -> Bool {
        patterns.isEmpty || anyMatchesAnyPattern(strs, patterns)
    }

    private func emptyOrNotAnyMatchesAnyPattern(_ strs: [String], _ patterns: [Regex]) -> Bool {
        patterns.isEmpty || !anyMatchesAnyPattern(strs, patterns)
    }

    private func emptyOrMatchesAnyString(_ str: String, _ strs: Set<String>) -> Bool {
        strs.isEmpty || strs.contains(str)
    }

    private func emptyOrNotMatchesAnyString(_ str: String, _ strs: Set<String>) -> Bool {
        strs.isEmpty || !strs.contains(str)
    }

    private func emptyOrMatchesAnyFileType(_ fileType: FileType, _ fileTypes: [FileType]) -> Bool {
        fileTypes.isEmpty || fileTypes.contains(fileType)
    }

    private func emptyOrNotMatchesAnyFileType(_ fileType: FileType, _ fileTypes: [FileType]) -> Bool {
        fileTypes.isEmpty || !fileTypes.contains(fileType)
    }

    public func isMatchingDirPathByHidden(_ dirPath: String) -> Bool {
        settings.includeHidden || !FileUtil.isHiddenPath(dirPath)
    }

    public func isMatchingDirPathByInPatterns(_ dirPath: String) -> Bool {
        emptyOrAnyMatchesAnyPattern(FileUtil.getPathComponents(dirPath), settings.inDirPatterns)
    }

    public func isMatchingDirPathByOutPatterns(_ dirPath: String) -> Bool {
        emptyOrNotAnyMatchesAnyPattern(FileUtil.getPathComponents(dirPath), settings.outDirPatterns)
    }

    public func isTraversableDirPath(_ dirPath: String) -> Bool {
        isMatchingDirPathByHidden(dirPath)
        && isMatchingDirPathByOutPatterns(dirPath)
    }

    public func isMatchingDirPath(_ dirPath: String) -> Bool {
        isMatchingDirPathByHidden(dirPath)
        && isMatchingDirPathByInPatterns(dirPath)
        && isMatchingDirPathByOutPatterns(dirPath)
    }

    public func isNullOrMatchingDirPath(_ dirPath: String?) -> Bool {
        dirPath == nil
        || (isMatchingDirPathByHidden(dirPath!)
        && isMatchingDirPathByInPatterns(dirPath!)
        && isMatchingDirPathByOutPatterns(dirPath!))
    }

    public func isMatchingFileNameByHidden(_ fileName: String) -> Bool {
        settings.includeHidden || !FileUtil.isHiddenName(fileName)
    }
    
    public func isMatchingArchiveExtension(_ ext: String) -> Bool {
        emptyOrMatchesAnyString(ext, settings.inArchiveExtensions)
        && emptyOrNotMatchesAnyString(ext, settings.outArchiveExtensions)
    }

    public func isMatchingArchiveExtensionForFilePath(_ filePath: String) -> Bool {
        if !settings.inArchiveExtensions.isEmpty || !settings.outArchiveExtensions.isEmpty {
            let ext = FileUtil.getExtension(filePath)
            return isMatchingArchiveExtension(ext)
        }
        return true
    }

    public func isMatchingArchiveFileName(_ fileName: String) -> Bool {
        emptyOrMatchesAnyPattern(fileName, settings.inArchiveFilePatterns)
        && emptyOrNotMatchesAnyPattern(fileName, settings.outArchiveFilePatterns)
    }

    public func isMatchingArchiveFileNameForFilePath(_ filePath: String) -> Bool {
        if !settings.inArchiveFilePatterns.isEmpty || !settings.outArchiveFilePatterns.isEmpty {
            let fileName = URL(fileURLWithPath: filePath).lastPathComponent
            return isMatchingArchiveFileName(fileName)
        }
        return true
    }

    public func isMatchingArchiveFilePath(_ filePath: String) -> Bool {
        isMatchingArchiveExtensionForFilePath(filePath)
        && isMatchingArchiveFileNameForFilePath(filePath)
    }

    public func isMatchingArchiveFileResult(_ fileResult: FileResult) -> Bool {
        isMatchingArchiveFilePath(fileResult.filePath)
    }


    public func isMatchingExtension(_ ext: String) -> Bool {
        emptyOrMatchesAnyString(ext, settings.inExtensions)
        && emptyOrNotMatchesAnyString(ext, settings.outExtensions)
    }

    public func isMatchingExtensionForFilePath(_ filePath: String) -> Bool {
        if !settings.inExtensions.isEmpty || !settings.outExtensions.isEmpty {
            let ext = FileUtil.getExtension(filePath)
            return isMatchingExtension(ext)
        }
        return true
    }

    public func isMatchingFileName(_ fileName: String) -> Bool {
        emptyOrMatchesAnyPattern(fileName, settings.inFilePatterns)
        && emptyOrNotMatchesAnyPattern(fileName, settings.outFilePatterns)
    }

    public func isMatchingFileNameForFilePath(_ filePath: String) -> Bool {
        if !settings.inFilePatterns.isEmpty || !settings.outFilePatterns.isEmpty {
            let fileName = URL(fileURLWithPath: filePath).lastPathComponent
            return isMatchingFileName(fileName)
        }
        return true
    }

    public func isMatchingFilePath(_ filePath: String) -> Bool {
        isMatchingExtensionForFilePath(filePath)
        && isMatchingFileNameForFilePath(filePath)
    }

    public func isMatchingFileType(_ fileType: FileType) -> Bool {
        emptyOrMatchesAnyFileType(fileType, settings.inFileTypes)
        && emptyOrNotMatchesAnyFileType(fileType, settings.outFileTypes)
    }

    public func isMatchingFileSize(_ fileSize: UInt64) -> Bool {
        (settings.maxSize <= 0 || fileSize <= settings.maxSize) &&
        (settings.minSize <= 0 || fileSize >= settings.minSize)
    }

    public func isMatchingLastMod(_ lastMod: Date?) -> Bool {
        (settings.maxLastMod == nil || lastMod! <= settings.maxLastMod!) &&
        (settings.minLastMod == nil || lastMod! >= settings.minLastMod!)
    }

    public func isMatchingFileResult(_ fileResult: FileResult) -> Bool {
        isMatchingFilePath(fileResult.filePath)
        && isMatchingFileType(fileResult.fileType)
        && isMatchingFileSize(fileResult.fileSize)
        && isMatchingLastMod(fileResult.lastMod)
    }

    public func filterArchiveFilePathToFileResult(_ filePath: String) -> FileResult? {
        if !settings.includeArchives && !settings.archivesOnly {
            return nil
        }

        if (!isMatchingArchiveFilePath(filePath)) {
            return nil
        }

        let fileSize: UInt64 = 0
        let lastMod: Date? = nil

        return FileResult(filePath: filePath, fileType: FileType.archive, fileSize: fileSize, lastMod: lastMod)
    }

    public func filterRegularFilePathToFileResult(_ filePath: String, _ fileType: FileType) -> FileResult? {
        if settings.archivesOnly {
            return nil
        }

        if !isMatchingFilePath(filePath) || !isMatchingFileType(fileType) {
            return nil
        }

        var fileSize: UInt64 = 0
        var lastMod: Date? = nil
        if self.settings.needSize() || self.settings.needLastMod() {
            do {
                let stat: [FileAttributeKey: Any] = try FileUtil.getFileAttributes(filePath)
                if self.settings.needSize() { fileSize = stat[FileAttributeKey.size] as! UInt64 }
                if self.settings.needLastMod() { lastMod = stat[FileAttributeKey.modificationDate] as? Date }

            } catch {
                print("Error: \(error)")
            }
        }

        if !isMatchingFileSize(fileSize) || !isMatchingLastMod(lastMod) {
            return nil
        }

        return FileResult(filePath: filePath, fileType: fileType, fileSize: fileSize, lastMod: lastMod)
    }

    public func filterToFileResult(_ filePath: String) -> FileResult? {
        let (parent, fileName) = FileUtil.splitPath(filePath)
        if !isNullOrMatchingDirPath(parent) || !isMatchingFileNameByHidden(fileName) {
            return nil
        }

        let fileType = fileTypes.getFileType(fileName)
        if fileType == FileType.archive {
            return filterArchiveFilePathToFileResult(filePath)
        }
        return filterRegularFilePathToFileResult(filePath, fileType)
    }

    // gets all FileResults recursively
    private func recGetFileResults(_ dirPath: String, minDepth: Int32, maxDepth: Int32, currentDepth: Int32) -> [FileResult] {
        var fileResults = [FileResult]()
        var recurse: Bool = true
        if currentDepth == maxDepth {
            recurse = false
        } else if (maxDepth > -1 && currentDepth > maxDepth) {
            return fileResults
        }

        let dirElems = try! FileManager.default.contentsOfDirectory(atPath: dirPath)
        var subDirs = [String]()

        for dirElem in dirElems {
            let subPath = FileUtil.joinPath(dirPath, childPath: dirElem)
            var linkIsDir: Bool = false
            var linkIsFile: Bool = false
            if FileUtil.isSymlink(subPath) {
                if settings.followSymlinks {
                    // Determine if dir or file
                    if let resolvedPath = FileUtil.getSymlinkTarget(subPath) {
                        if FileUtil.isDirectory(resolvedPath) {
                            linkIsDir = true
                        } else if FileUtil.isReadableFile(subPath) {
                            linkIsFile = true
                        }
                    }
                } else {
                    continue
                }
            }
            if FileUtil.isDirectory(subPath) || linkIsDir {
                if recurse && isTraversableDirPath(dirElem) {
                    subDirs.append(subPath)
                }
            } else if (FileUtil.isReadableFile(subPath) || linkIsFile) && (minDepth < 0 || currentDepth >= minDepth) {
                let fileResult = filterToFileResult(subPath)
                if fileResult != nil {
                    fileResults.append(fileResult!)
                }
            }
        }

        for subDir in subDirs {
            let dirResults = recGetFileResults(subDir, minDepth: minDepth, maxDepth: maxDepth, currentDepth: (currentDepth + 1))
            fileResults.append(contentsOf: dirResults)
        }

        return fileResults
    }

    // gets all FileResults recursively
    private func getFileResults(_ path: String) throws -> [FileResult] {
        var p = path
        if !FileUtil.exists(path) {
            p = FileUtil.expandPath(path)
            if !FileUtil.exists(p) {
                throw FindError(msg: STARTPATH_NOT_FOUND)
            }
        }
        var fileResults = [FileResult]()
        var linkIsDir: Bool = false
        var linkIsFile: Bool = false
        if FileUtil.isSymlink(p) {
            if settings.followSymlinks {
                // Determine if dir or file
                if let resolvedPath = FileUtil.getSymlinkTarget(p) {
                    if FileUtil.isDirectory(resolvedPath) {
                        linkIsDir = true
                    } else if FileUtil.isReadableFile(p) {
                        linkIsFile = true
                    }
                }
            } else {
                throw FindError(msg: STARTPATH_NOT_MATCH_FIND_SETTINGS)
            }
        }
        if FileUtil.isDirectory(p) || linkIsDir {
            // if maxDepth is zero, we can skip since a directory cannot be a result
            if settings.maxDepth == 0 {
                return fileResults
            }
            if self.isTraversableDirPath(p) {
                let maxDepth = settings.recursive ? settings.maxDepth : 1
                let pathResults = recGetFileResults(p, minDepth: settings.minDepth, maxDepth: maxDepth, currentDepth: 1)
                fileResults.append(contentsOf: pathResults)
            } else {
                throw FindError(msg: STARTPATH_NOT_MATCH_FIND_SETTINGS)
            }
        } else if FileUtil.isReadableFile(p) || linkIsFile {
            // if minDepth > zero, we can skip since the file is at depth zero
            if settings.minDepth > 0 {
                return fileResults
            }
            if let fileResult = filterToFileResult(p) {
                fileResults.append(fileResult)
            } else {
                throw FindError(msg: STARTPATH_NOT_MATCH_FIND_SETTINGS)
            }
        } else {
            throw FindError(msg: STARTPATH_NOT_MATCH_FIND_SETTINGS)
        }
        return fileResults
    }

    public func find() throws -> [FileResult] {
        var fileResults = [FileResult]()
        for p in settings.paths {
            let pathResults: [FileResult] = try getFileResults(p)
            fileResults.append(contentsOf: pathResults)
        }
        if fileResults.count > 1 {
            let fileResultSorter = FileResultSorter(settings: settings)
            return fileResultSorter.sort(fileResults)
        }
        return fileResults
    }

    func getMatchingDirs(_ fileResults: [FileResult]) -> [String] {
        fileResults.map {
            URL(fileURLWithPath: $0.filePath).deletingLastPathComponent().path
        }.sorted().unique()
    }

    public func printMatchingDirs(_ fileResults: [FileResult], _ formatter: FileResultFormatter) -> Void {
        let dirs = getMatchingDirs(fileResults)
        if dirs.isEmpty {
            logMsg("\nMatching directories: 0")
        } else {
            logMsg("\nMatching directories (\(dirs.count)):")
            for dir in dirs {
                logMsg(formatter.formatDirPath(dir))
            }
        }
    }

    func getMatchingFiles(_ fileResults: [FileResult]) -> [String] {
        fileResults.map(\.filePath)
    }

    public func printMatchingFiles(_ fileResults: [FileResult], _ formatter: FileResultFormatter) -> Void {
        if fileResults.isEmpty {
            logMsg("\nMatching files: 0")
        } else {
            logMsg("\nMatching files (\(fileResults.count)):")
            for fileResult in fileResults {
                logMsg(formatter.formatFileResult(fileResult))
            }
        }
    }
}
